File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.stats import MatchPlayerStatsOut


class PlayerMatchPerformanceRow(BaseModel):
    match_id: int
    match_date: datetime
    opponent: str
    competition: str | None
    status: str
    stats: MatchPlayerStatsOut
    rating: int | None = None
    note: str | None = None
    # From pass events: +1 / −1 clicks in live recording (successful vs corrected/removed)
    passes_success: int = 0
    passes_unsuccess: int = 0


class LeaderboardRow(BaseModel):
    player_id: int
    first_name: str
    last_name: str
    jersey_number: int
    games: int
    goals: int
    assists: int
    passes: int
    yellow_cards: int
    red_cards: int
    avg_rating: float | None = None


class TeamMatchBreakdownRow(BaseModel):
    """Per-match team totals (sum over all players) for charts and trends."""

    match_id: int
    match_date: datetime
    opponent: str
    competition: str | None
    status: str
    goals: int
    assists: int
    passes: int
    passes_success: int
    passes_unsuccess: int
    shots_on_goal: int
    shots_off_goal: int
    won_duels: int
    lost_duels: int
    fouls: int
    yellow_cards: int
    red_cards: int

    class Config:
        from_attributes = True

File: frontend/src/api/analytics.ts
import { api } from "./client";
import type { PlayerStats } from "./types";

export type PlayerPerformanceRow = {
  match_id: number;
  match_date: string;
  opponent: string;
  competition?: string | null;
  status: string;
  stats: { match_id: number; player_id: number } & PlayerStats;
  rating?: number | null;
  note?: string | null;
  /** Pass events with +1 (recorded) */
  passes_success: number;
  /** Pass events with −1 (correction / removed) */
  passes_unsuccess: number;
};

export type LeaderboardRow = {
  player_id: number;
  first_name: string;
  last_name: string;
  jersey_number: number;
  games: number;
  goals: number;
  assists: number;
  passes: number;
  yellow_cards: number;
  red_cards: number;
  avg_rating?: number | null;
};

export async function playerPerformance(playerId: number, seasonId?: number) {
  const res = await api.get<PlayerPerformanceRow[]>(`/players/${playerId}/performance`, {
    params: seasonId ? { season_id: seasonId } : undefined,
  });
  return res.data;
}

export async function seasonLeaderboards(seasonId: number) {
  const res = await api.get<LeaderboardRow[]>(`/seasons/${seasonId}/leaderboards`);
  return res.data;
}

export type TeamMatchBreakdownRow = {
  match_id: number;
  match_date: string;
  opponent: string;
  competition?: string | null;
  status: string;
  goals: number;
  assists: number;
  passes: number;
  passes_success: number;
  passes_unsuccess: number;
  shots_on_goal: number;
  shots_off_goal: number;
  won_duels: number;
  lost_duels: number;
  fouls: number;
  yellow_cards: number;
  red_cards: number;
};

export async function teamMatchesBreakdown(seasonId: number) {
  const res = await api.get<TeamMatchBreakdownRow[]>(
    `/seasons/${seasonId}/team-matches-breakdown`
  );
  return res.data;
}

